src/ocean/task/util/Timer.d

```diff
@@ -124,6 +124,151 @@ unittest
 
 /*******************************************************************************
 
+    Similar to `theScheduler.await` but also has waiting timeout. Calling task
+    will be resumed either if awaited task finished or timeout is hit, whichever
+    happens first.
+
+    Params:
+        task = task to await for
+        micro_seconds = timeout duration
+
+    Returns:
+        'true' if resumed via timeout, 'false' otherwise
+
+*******************************************************************************/
+
+public bool awaitOrTimeout ( Task task, uint micro_seconds )
+{
+    auto context = Task.getThis();
+    assert (context !is null);
+
+    if (.timer is null)
+        .timer = new typeof(timer);
+
+    auto scheduled_event = .timer.schedule(
+        // EventData setup is run from the same fiber so it is ok to reference
+        // variable from this function stack
+        ( ref EventData event )
+        {
+            event.to_resume = context;
+        },
+        // Callback of fired timer is run from epoll context and here it is
+        // only legal to use data captured as EventData field (or other heap
+        // allocated data)
+        ( ref EventData event )
+        {
+            debug_trace("Resuming task <{}> because of await timeout",
+                cast(void*) event.to_resume);
+            event.to_resume.resume();
+        },
+        micro_seconds
+    );
+
+    task.terminationHook(&context.resume);
+    task.terminationHook(&scheduled_event.unregister);
+
+    // force async scheduling to avoid checking if this context needs
+    // suspend/resume and do it unconditionally
+    theScheduler.queue(task);
+    context.suspend();
+    // skip one scheduler event loop cycle to allow awaited task to finish with
+    // all termination hooks
+    theScheduler.processEvents();
```

Comment:
Haha, yes of course :-)

Reply:
Actually no, this is not good enough - the same fiber could be busy with other task by that time. I will update to provide required info by task itself.